muxers/yamux/src/lib.rs

```diff
@@ -63,7 +63,11 @@ pub struct Muxer<S> {
     inbound_stream_waker: Option<Waker>,
 }
 
-const MAX_BUFFERED_INBOUND_STREAMS: usize = 25;
+/// Yamux does not have any built-in backpressure mechanism.
+///
+/// Implementations are encouraged to not open more than 256 unacknowledged streams, see <https://github.com/libp2p/specs/tree/master/yamux#ack-backlog>.
+/// Thus, it makes sense to buffer up to 256 streams before dropping them for good interoperability.
+const MAX_BUFFERED_INBOUND_STREAMS: usize = 256;
```

Comment:
I am not sure I follow the reasoning above. At the point where a new stream is retrieved through `this.poll_inner`, the stream will be acknowledged to the remote, thus allowing the remote to open another new stream.

As far as I can tell, bumping this limit just delays running into the limit itself, no?

Reply:
Thanks for questioning this!

> At the point where a new stream is retrieved through `this.poll_inner`, the stream will be acknowledged to the remote

I don't think this is true. We emit streams in `rust-yamux` without acknowledging them directly because we don't want to send a frame for _just_ acknowledging a stream. I'll have to dig into this in detail but I think we delay the acknowledgement until we send the first payload on this stream.

The first payload will be `multistream-select` which in our case happens directly after we have removed a stream from this buffer. Meaning if our `Connection` is slow in removing streams from this buffer, the remote should eventually stop opening new streams.

I think it makes sense to write a test against this just to be sure but for that we land the ACK backlog changes first: https://github.com/libp2p/rust-yamux/issues/150

Reply:
If the `WindowUpdateMode` is set to `OnRead` (which is the default), then we are not sending a `Frame` on a newly opened inbound stream until the user first uses it. To acknowledge the stream in that case, we preemptively set the `ACK` flag:

https://github.com/libp2p/rust-yamux/blob/72ccd5734f71971fc1f02ac4a5f43b8eb4dff660/yamux/src/connection.rs#L656

Let me know if you see anything wrong with the reasoning here.

Reply:
I realized I still had a fully-working branch for the ACK backlog locally so I just quickly added a test for this as well and it confirmed my intuition: https://github.com/libp2p/rust-yamux/pull/153